File: micro/locus_instantiation.c
// C equivalent of locus_instantiation.hl.
// Helper fn heap-allocates Empty, calls a method to return a field,
// XORs the result into a sink — matches the Hale/Go bench shape so
// nothing gets DCE'd. getpid() seeds the values. Caveats: C frees
// each instance immediately where Go defers to GC; and the asm
// barrier keeps clang from eliding the malloc/free pair the way
// Go's runtime can't.
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>

static long long now_ns(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (long long)ts.tv_sec * 1000000000LL + (long long)ts.tv_nsec;
}

typedef struct {
    long long v;
} Empty;

__attribute__((noinline)) static long long empty_read(Empty *e) { return e->v; }

__attribute__((noinline)) static long long instantiate_one(long long seed) {
    Empty *e = malloc(sizeof *e);
    if (!e) exit(1);
    e->v = seed;
    asm volatile("" :: "r"(e) : "memory"); // keep the heap allocation real
    long long r = empty_read(e);
    free(e);
    return r;
}

int main(void) {
    long long iters = 100000;
    long long pid = getpid();
    long long t0 = now_ns();
    long long sink = 0;
    for (long long i = 0; i < iters; i++) {
        sink ^= instantiate_one(i + pid);
    }
    long long elapsed = now_ns() - t0;
    printf("iters=%lld\n", iters);
    printf("sink=%lld\n", sink);
    printf("elapsed_ns=%lld\n", elapsed);
    return 0;
}

File: micro/locus_instantiation.rs
// Rust equivalent of locus_instantiation.hl.
// Helper fn allocates Empty on the heap (Box), calls a method to
// return a field, XORs the result into a sink. Matches the Hale
// bench's shape so neither side gets DCE'd. black_box pins the Box —
// unlike Go's escape analysis, LLVM would otherwise stack-promote or
// elide the allocation entirely, and the point is heap-per-instance.

use std::hint::black_box;
use std::time::Instant;

struct Empty {
    v: i64,
}

impl Empty {
    #[inline(never)]
    fn read(&self) -> i64 {
        self.v
    }
}

#[inline(never)]
fn instantiate_one(seed: i64) -> i64 {
    let e = black_box(Box::new(Empty { v: seed }));
    e.read()
}

fn main() {
    let iters: i64 = 100000;
    let pid = std::process::id() as i64;
    let t0 = Instant::now();
    let mut sink: i64 = 0;
    for i in 0..iters {
        sink ^= instantiate_one(i + pid);
    }
    let elapsed = t0.elapsed().as_nanos();
    println!("iters={}", iters);
    println!("sink={}", sink);
    println!("elapsed_ns={}", elapsed);
}
